src/url.rs

```diff
@@ -314,6 +358,120 @@ impl PyMultiHostUrl {
     fn __getnewargs__(&self) -> (String,) {
         (self.__str__(),)
     }
+
+    #[classmethod]
+    #[pyo3(signature=(*, scheme, hosts=None, path=None, query=None, fragment=None, host=None, user=None, password=None, port=None))]
+    #[allow(clippy::too_many_arguments)]
+    pub fn build<'a>(
+        cls: &'a PyType,
+        scheme: &str,
+        hosts: Option<Vec<MultiHostUrlHost>>,
+        path: Option<&str>,
+        query: Option<&str>,
+        fragment: Option<&str>,
+        // convenience parameters to build with a single host
+        host: Option<&str>,
+        user: Option<&str>,
+        password: Option<&str>,
+        port: Option<&str>,
+    ) -> PyResult<&'a PyAny> {
+        let mut url = if hosts.is_some() && (host.is_some() || user.is_some() || password.is_some() || port.is_some()) {
+            return Err(PyValueError::new_err(
+                "expected one of `hosts` or singular values to be set.",
+            ));
+        } else if let Some(hosts) = hosts {
+            // check all of host / user / password / port empty
+            // build multi-host url
+            let mut multi_url = format!("{scheme}://");
+            for (index, single_host) in hosts.iter().enumerate() {
+                if single_host.is_empty() {
+                    return Err(PyValueError::new_err(
+                        "expected one of 'host', 'username', 'password' or 'port' to be set",
+                    ));
+                }
+                multi_url.push_str(&single_host.to_string());
+                if index != hosts.len() - 1 {
+                    multi_url.push(',');
+                };
+            }
+            multi_url
+        } else if let Some(host) = host {
+            let user_password = match (user, password) {
+                (Some(user), None) => format!("{user}@"),
+                (None, Some(password)) => format!(":{password}@"),
+                (Some(user), Some(password)) => format!("{user}:{password}@"),
+                (None, None) => String::new(),
+            };
+            let mut single_host_url = format!("{scheme}://{user_password}{host}");
+            if let Some(port) = port {
+                single_host_url.push(':');
+                single_host_url.push_str(port);
+            };
+            single_host_url
```

Comment:
I wonder if here it makes sense to create a `MultiHostUrlHost` and then use that?

```suggestion
            let url_host = MultiHostUrlHost {
                username: username.map(Into::into),
                password: password.map(Into::into),
                host: host.map(Into::into),                
                port: port.map(Into::into),
            };
            format!("{scheme}://{url_host}")
```

I notice the logic in `MultiHostUrlHost` is different to the `match` here, perhaps that is deliberate?